### PR TITLE
Enable defines

### PR DESCRIPTION
**Pull request by mstamer**
Easiest way to review changes is as a whole, not each commit...  Change mainly combines the if/else for the openthred-config .h header into a single inclusion.  Also adds missing define to dhcp6_client.cpp.   Sorry no squashing, too unfamiliar with git, squash command caused too many conflicts...

### CODE REVIEW

**Comment by darconeous** [comment]
So what problem are we trying to solve here?

**Comment by mstamer** [comment]
Hello, the main issue was that the #if OPENTHREAD_ENABLE_DHCP6_CLIENT was missing from dhcp6_client.cpp.  At the same time, it was requested to change the if/else/endif scheme into just a simple file inclusion...  So the file openthread-enable-defines.h is also new.
Basically change:

`#ifdef OPENTHREAD_CONFIG_FILE`
`#include OPENTHREAD_CONFIG_FILE`
`#else`
`#include <openthread-config.h>`
`#endif`

Into:

`#include <openthread-enable-defines.h>`

This would be included in all core source files, and any other files that had this if/else/endif scheme...

**Comment by darconeous** [comment]
> it was requested to change the if/else/endif scheme into just a simple file inclusion

Requested by whom?

**Comment by DuaneEllis-TI** [comment]
I do agree that moving these to a single/common file is a better choice

Rather then LOTS of  #if/#else/#endif for the "config.h" file having a single one is better.

That said, being able to pass a different FILENAME on the command line is required.
that appears to remain.

Need to look this over more.

File: src/core/api/crypto_api.cpp [hunk 1]
```diff
@@ -31,8 +31,10 @@
  *   This file implements the OpenThread Crypto API.
  */
 
+#include <openthread/openthread_enable_defines.h>
 #include <openthread/crypto.h>
 
+
```
Comment (aeliot) [inline src/core/api/crypto_api.cpp]: nit: extra newline

File: src/core/meshcop/energy_scan_client.cpp [hunk 2]
```diff
@@ -32,12 +32,11 @@
  */
 
 #define WPP_NAME "energy_scan_client.tmh"
+#include <openthread/openthread_enable_defines.h>
+
+
+
```
Comment (aeliot) [inline src/core/meshcop/energy_scan_client.cpp]: nit: extra newlines

File: src/core/meshcop/panid_query_client.cpp [hunk 3]
```diff
@@ -33,14 +33,11 @@
...
-#ifdef OPENTHREAD_CONFIG_FILE
-#include OPENTHREAD_CONFIG_FILE
-#else
-#include <openthread-config.h>
-#endif
-
+#include <openthread/openthread_enable_defines.h>
 #include "panid_query_client.hpp"
 
+
```
Comment (aeliot) [inline src/core/meshcop/panid_query_client.cpp]: nit: extra newlines

File: src/core/net/dhcp6_client.cpp [hunk 4]
```diff
@@ -51,6 +47,9 @@
 #include "net/dhcp6.hpp"
 #include "thread/thread_netif.hpp"
 
+
```
Comment (aeliot) [inline src/core/net/dhcp6_client.cpp]: nit: extra newline

File: src/core/net/dhcp6_client.cpp [hunk 5]
```diff
@@ -725,3 +724,6 @@ otError Dhcp6Client::ProcessIaAddress(Message &aMessage, uint16_t aOffset)
 
 }  // namespace Dhcp6
 }  // namespace ot
+
+#endif //OPENTHREAD_ENABLE_DHCP6_CLIENT
```
Comment (aeliot) [inline src/core/net/dhcp6_client.cpp]: nit: space after `//`

File: src/core/thread/data_poll_manager.cpp [hunk 6]
```diff
@@ -32,15 +32,16 @@
  */
 
 #define WPP_NAME "data_poll_manager.tmh"
+#include <openthread/openthread_enable_defines.h>
+
```
Comment (aeliot) [inline src/core/thread/data_poll_manager.cpp]: nit: extra newlines

File: src/core/thread/data_poll_manager.cpp [hunk 7]
```diff
@@ -32,15 +32,16 @@
...
 
-#ifdef OPENTHREAD_CONFIG_FILE
-#include OPENTHREAD_CONFIG_FILE
-#else
-#include <openthread-config.h>
-#endif
 
 #include "data_poll_manager.hpp"
 
+
```
Comment (aeliot) [inline src/core/thread/data_poll_manager.cpp]: nit: extra newlines

File: src/ncp/ncp_buffer.hpp [hunk 8]
```diff
@@ -381,36 +382,36 @@ class NcpFrameBuffer
     otError         OutFramePrepareMessage(void);
     otError         OutFrameFillMessageBuffer(void);
 
-    uint8_t * const  mBuffer;                    // Pointer to the buffer used to store the data.
-    uint8_t * const  mBufferEnd;                 // Points to after the end of buffer.
-    const uint16_t   mBufferLength;              // Length of the the buffer.
+    uint8_t * const mBuffer;                    // Pointer to the buffer used to store the data.
```
Comment (aeliot) [inline src/ncp/ncp_buffer.hpp]: It looks like you changed the alignment for some of these lines, but not all.  I'm not sure why this change would be needed, the original alignment looks correct.

File: include/openthread/openthread_enable_defines.h [hunk 9]
```diff
@@ -0,0 +1,48 @@
+/*
+ *  Copyright (c) 2017, The OpenThread Authors.
```
Comment (DuaneEllis-TI) [inline include/openthread/openthread_enable_defines.h]: Maybe we should rename this file, I think this file should be named: openthread_config.h

My objection here is the "enables define" name just seems weird/out of place.

the *GENERATED* file should be come:  openthread-config-generated.h

Then platform that need to use a specific configuration can provide their own

Something like this:
```
#if  !defined(OPENTHREAD_CONFIG_FILE)
#define OPENTHREAD_CONFIG_FILE  "openthread-config-generated.h"
#endif

#include OPENTHREAD_CONFIG_FILE
```

Then, as platforms

**Comment by abtink** [comment]
@mstamer, thanks for addressing this. I am very much in favor of this (was considering doing this actually).

I have a proposal/suggestion regarding this.

We have two different header files where the settings/configuration options can be defined, the `openthread-config.h` which may be generated by autoconf/`configure.ac` and then we can have a project specific configuration header file given by `OPENTHREAD_PROJECT_CORE_CONFIG_FILE` (with default settings in `"openthread-core-default-config.h"`). We need to ensure all Openthread source files include both. 

We already have a private include file named, `"openthread-core-config.h"` (where we have the #ifdef for the project core config).  My suggestion is to that we update this file and use this as the main setting/configuration related header file for internal OpenThread use. Basically we can add the other #ifdef for `openthread-config.h` in the same file `"openthread-core-config.h"`:
```
 #define OPENTHREAD_CORE_CONFIG_H_IN
 
+#ifdef OPENTHREAD_CONFIG_FILE
+#include OPENTHREAD_CONFIG_FILE
+#else
+#include <openthread-config.h>
+#endif
+
 #ifdef OPENTHREAD_PROJECT_CORE_CONFIG_FILE
 #include OPENTHREAD_PROJECT_CORE_CONFIG_FILE
 #endif

```
We can then include the `"openthread-core-config.h"` at the top of every private/local `hpp` file. Since every `cpp` file includes its corresponding `hpp` file before any other header file, we always have the config/setting header file before any other code everywhere. I feel this makes the internal use much simpler and less error-prone.
For OpenThread public C header files (under `include\`), for now we can continue using the current/existing model.

Thoughts?

File: src/ncp/ncp_buffer.hpp (hunk 8 above)
Comment (mstamer) [inline src/ncp/ncp_buffer.hpp]: This was automatically changed with the "make pretty" prior to committing...  I can undo if needed...

Comment (aeliot) [inline src/ncp/ncp_buffer.hpp]: Thats odd, if I run ``make pretty`` on a clean version of master, I dont see it trying to change this file.  I think these lines should be reverted.  Nothing actually changed on these lines, and the formatting didn't even change all the lines, so its created some skew in the indentations.
We are slowly switching from astyle to uncrustify anyways.

**Comment by mstamer** [comment]
Regarding CLA, it should be signed now via corporate.  Under S&C ELectric.  Please let me know if any issues on the license front...

**Comment by jwhui** [comment]
@mstamer, thanks for the update on the CLA.  The internal status currently shows as "Pending".  I'll let you know when it is "Accepted".

**Comment by mstamer** [comment]
Going over all the comments, I see the below as my "to-do" list.  (though some items are still under discussion)...

1. remove any extra new lines and white space - keep it clean (use diff to find all)
2. remove the "make pretty" changes my build did to ncp_buffer.hpp - figured out this was due to incorrect astyle version on my system (astyle 3.0 - corrected to use the one pulled from tools).

Under discussion:

Either:
1. rename openthread_enable_defines.h to openthread_config.h
2. rename generated openthread-config.h to openthread-config-generated.h
3. update "new" openthread-config.h to use new #if/#define/#endif/#include scheme 

Or
1. Use openthread-core-config.h as main header by adding the new if/define/endif/include scheme to the beginning of this file.
2. Include this openthread-core-config.h at the top of every private/local header file.
3. OpenThread public C header files (under include\), for now we can continue using the current/existing model (if/else/endif in header file as required)

Or combination of the two:
1. rename openthread_enable_defines.h to openthread_config.h
2. rename generated openthread-config.h to openthread-config-generated.h
3. update "new" openthread-config.h to use new #if/#define/#endif/#include scheme 
4. Use openthread-core-config.h as main header by adding the "new" incldue openthread_config.h to the beginning of this file.
5. Include this openthread-core-config.h at the top of every private/local header file.
6. Include "new" openthread-config.h in the OpenThread public C header files (under include\) where required.

Or some other variation not yet discussed...  

I vote for the combination approach...  private local files will include openthread-core-config.h, and the public headers will include openthread-config.h

After going over the code, I did notice I missed the following files with this update (in the /include directory).  
I will update them as well once we finalize the approach to move forward with.
\include\openthread\jam_detection.h
\include\openthread\dns.h
\include\openthread\coap.h
\include\openthread\child_supervision.h
\include\openthread\border_agent_proxy.h

thoughts?

**Comment by mstamer** [comment]
One more question, is there any style guide regarding the spacing of include files?  Should each include be separated by one newline?  or none?  or grouped by using <> vs ""?  As I am cleaning up, I see some have no newlines, where others have one, and others have some?   I did not see an answer to this on the posted style guide...  I'm tempted to remove all newlines between the includes...

**Comment by abtink** [comment]
👍  for the combination

**Comment by aeliot** [comment]
For the ``#include`` statements, I dont know if we have an official policy.  Generally, similar files are grouped together.  So you may have a block of standard includes, and a block for OpenThread includes.  Whatever seams like a logical grouping.  I would say that having ``\n\n`` between every ``#include`` is definitely too much though.

**Comment by jwhui** [comment]
The style guide does not currently mention this, but as @aeliot mentioned, the `#include` statements are typically grouped by the categories mentioned in the style guide:

- C++ Standard Library headers
- C Standard Library headers
- Third-party library headers
- First-party library headers
- Private or local headers

File: include/openthread/openthread_config.h [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+/**
+ * @file
+ * @brief
+ *  This file includes required defines config header.
+ */
+
+#ifndef OPENTHREAD_CONFIG_H_
+#define OPENTHREAD_CONFIG_H_
+
+#if  !defined(OPENTHREAD_CONFIG_FILE)
```
Comment (aeliot) [inline include/openthread/openthread_config.h]: nit: There is an extra space between if and !

File: src/core/api/thread_ftd_api.cpp [hunk 11]
```diff
@@ -33,16 +33,14 @@
...
-#else
-#include <openthread-config.h>
-#endif
+#include <openthread/openthread_config.h>
 
 #if OPENTHREAD_FTD
 
-#include <openthread/thread_ftd.h>
 
+
```
Comment (aeliot) [inline src/core/api/thread_ftd_api.cpp]: nit: extra newlines

File: src/core/thread/data_poll_manager.cpp [hunk 12]
```diff
@@ -32,15 +32,16 @@
  */
 
 #define WPP_NAME "data_poll_manager.tmh"
+#include <openthread/openthread_config.h>
+
```
Comment (aeliot) [inline src/core/thread/data_poll_manager.cpp]: nit: extra newlines

File: src/core/thread/energy_scan_server.cpp [hunk 13]
```diff
@@ -32,15 +32,14 @@
...
-#ifdef OPENTHREAD_CONFIG_FILE
-#include OPENTHREAD_CONFIG_FILE
-#else
-#include <openthread-config.h>
-#endif
 
 #include "energy_scan_server.hpp"
 
+
+
```
Comment (aeliot) [inline src/core/thread/energy_scan_server.cpp]: nit: extra newlines

File: src/core/thread/network_data_leader_ftd.cpp [hunk 14]
```diff
@@ -33,13 +33,10 @@
 
 #if OPENTHREAD_FTD
 
+
```
Comment (aeliot) [inline src/core/thread/network_data_leader_ftd.cpp]: nit: extra newline

**Comment by jwhui** [comment]
@mstamer, can you resolve the merge conflicts?  This is also the reason why the build checks did not run.

**Comment by abtink** [comment]
(Sorry for closing it accidentally... wrong button...)

**Comment by mstamer** [comment]
Conflicts resolved and building...  Never got to the combination plan to include the openthread_config into the core_config for the core directory...  I guess some secondary ticket can will have to take care of that, if needed...
